src/lib/comms/sol-connman-impl-connman.c

```diff
@@ -0,0 +1,910 @@
+{
+    SOL_NULL_CHECK(service, NULL);
+
+    return service->name;
+}
+
+SOL_API const char *
+sol_connman_service_get_type(const struct sol_connman_service *service)
+{
+    SOL_NULL_CHECK(service, NULL);
+
+    return service->type;
+}
+
+SOL_API enum sol_connman_service_state
+sol_connman_service_get_state(const struct sol_connman_service *service)
+{
+    SOL_NULL_CHECK(service, SOL_CONNMAN_SERVICE_STATE_UNKNOWN);
+
+    return service->state;
+}
+
+SOL_API struct sol_network_link_addr *
+sol_connman_service_get_network_address(const struct sol_connman_service *service,
+    enum sol_network_family family)
+{
+    SOL_NULL_CHECK(service, NULL);
+
+    if (family == SOL_NETWORK_FAMILY_INET)
+        return service->link_addr;
+    else if (family == SOL_NETWORK_FAMILY_INET6)
+        return service->link_addr6;
+    else
+        return NULL;
+}
+
+SOL_API int32_t
+sol_connman_service_get_strength(const struct sol_connman_service *service)
+{
+    SOL_NULL_CHECK(service, -EINVAL);
+
+    return service->strength;
+}
+
+static int
+get_service_ip(sd_bus_message *m, struct sol_network_link_addr *link_addr)
+{
+    char *str;
+    int r;
+
+    SOL_NULL_CHECK(link_addr, -EINVAL);
+
+    r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "{sv}");
+    SOL_INT_CHECK(r, < 0, r);
+
+    do {
+        r = sd_bus_message_enter_container(m, SD_BUS_TYPE_DICT_ENTRY, "sv");
+        SOL_INT_CHECK_GOTO(r, < 1, end);
+
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &str);
+        SOL_INT_CHECK(r, < 0, r);
+        if (strcmp(str, "Address") == 0) {
+            char *address;
+
+            r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "s");
+            SOL_INT_CHECK(r, < 0, r);
+
+            r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &address);
+            SOL_INT_CHECK(r, < 0, r);
+
+            sol_network_link_addr_from_str(link_addr, address);
+
+            r = sd_bus_message_exit_container(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else {
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+
+        r = sd_bus_message_exit_container(m);
+        SOL_INT_CHECK(r, < 0, r);
+    } while (1);
+
+end:
+    if (r == 0)
+        r = sd_bus_message_exit_container(m);
+
+    return r;
+}
+
+static void
+get_service_ipv4(sd_bus_message *m, struct sol_connman_service *service)
+{
+    if (!service)
+        return;
+
+    if (!service->link_addr) {
+        service->link_addr = malloc(sizeof(*service->link_addr));
+        if (!service->link_addr)
+            return;
+        service->link_addr->family = SOL_NETWORK_FAMILY_INET;
+    }
+    get_service_ip(m, service->link_addr);
+}
+
+static void
+get_service_ipv6(sd_bus_message *m, struct sol_connman_service *service)
+{
+    if (!service)
+        return;
+
+    if (!service->link_addr6) {
+        service->link_addr6 = malloc(sizeof(*service->link_addr6));
```

Comment:
same here, no need to malloc. And if you convert to keep a sol_network_link, it's basically append to the vector (that will malloc for you :-D)